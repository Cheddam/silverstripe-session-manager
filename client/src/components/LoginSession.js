/* global window */
import React, { Component } from 'react';
import confirm from '@silverstripe/reactstrap-confirm';
import Config from 'lib/Config'; // eslint-disable-line
import backend from 'lib/Backend';
import moment from 'moment';
import i18n from 'i18n';
import PropTypes from 'prop-types';

class LoginSession extends Component {
    constructor(props) {
        super(props);

        this.state = {
            complete: false,
            failed: false,
            submitting: false,
        };

        this.logOut = this.logOut.bind(this);
    }

    async logOut() {
        // Confirm with the user
        const confirmMessage = i18n._t(
            'SessionManager.DELETE_CONFIRMATION',
            'Are you sure you want to delete this login session?'
        );
        const confirmTitle = i18n._t(
            'SessionManager.CONFIRMATION_TITLE',
            'Are you sure?'
        );
        const buttonLabel = i18n._t(
            'SessionManager.DELETE_CONFIRMATION_BUTTON',
            'Remove login session'
        );

        if (!await confirm(confirmMessage, { title: confirmTitle, confirmLabel: buttonLabel })) {
            return;
        }

        this.setState({ submitting: true });

        const endpoint = backend.createEndpointFetcher({
            url: `${this.props.LogOutEndpoint}/:id`,
            method: 'delete',
            payloadSchema: {
                id: { urlReplacement: ':id', remove: true }
            },
            defaultData: {
                SecurityID: Config.get('SecurityID')
            }
        });

        endpoint({
            id: this.props.ID
        })
            .then(response => response.json())
            .then(output => {
                const failed = !!output.error;
                this.setState({ complete: true, failed, submitting: false });
            })
            .catch(() => {
                this.setState({ complete: true, failed: true, submitting: false });
            });
    }

    render() {
        const lastAccessedElapsed = moment.utc(this.props.LastAccessed).fromNow();

        if (this.state.complete || this.state.submitting) {
            return null;
        }

        return (
          <div className="login-session">
            <div>{this.props.UserAgent}</div>
            {this.props.IsCurrent && <strong className={'text-success'}>{i18n._t(
                    'SessionManager.CURRENT',
                    'Current'
                )}</strong>}
            <div className="text-muted">
              {this.props.IPAddress}
              {!this.props.IsCurrent && `, ${i18n._t(
                        'SessionManager.LAST_ACTIVE',
                        'last active'
                    )} ${lastAccessedElapsed}`}
            </div>
            {!this.props.IsCurrent && <a
              href="javascript:void(0);" // eslint-disable-line
              onClick={this.logOut}
            >{i18n._t(
                'SessionManager.LOG_OUT',
                    'Log Out'
                )}</a>}
          </div>
        );
    }
}

const LoginSessionShape = PropTypes.shape({
    ID: PropTypes.number.isRequired,
    IPAddress: PropTypes.string,
    IsCurrent: PropTypes.bool,
    UserAgent: PropTypes.string,
    Persistent: PropTypes.number,
    Member: PropTypes.object,
    LastAccessed: PropTypes.string,
    LogOutEndpoint: PropTypes.string,
});

LoginSession.propTypes = LoginSessionShape;

LoginSession.defaultProps = {};

export default LoginSession;

export { LoginSessionShape };
