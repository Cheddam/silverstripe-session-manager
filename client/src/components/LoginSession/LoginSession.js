/* global window */
import React, { useState } from 'react';
import confirm from '@silverstripe/reactstrap-confirm';
import Config from 'lib/Config'; // eslint-disable-line
import backend from 'lib/Backend';
import moment from 'moment';
import i18n from 'i18n';
import PropTypes from 'prop-types';

function LoginSession(props) {
    const [loading, setLoading] = useState({ complete: false, failed: false, submitting: false });

    async function logOut() {
        // Confirm with the user
        const confirmMessage = i18n._t(
            'SessionManager.DELETE_CONFIRMATION',
            'Are you sure you want to delete this login session?'
        );
        const confirmTitle = i18n._t('SessionManager.CONFIRMATION_TITLE', 'Are you sure?');
        const buttonLabel = i18n._t('SessionManager.DELETE_CONFIRMATION_BUTTON', 'Remove login session');

        if (!await confirm(confirmMessage, { title: confirmTitle, confirmLabel: buttonLabel })) {
            return;
        }

        setLoading({ ...loading, submitting: true });

        const endpoint = backend.createEndpointFetcher({
            url: `${props.LogOutEndpoint}/:id`,
            method: 'delete',
            payloadSchema: {
                id: { urlReplacement: ':id', remove: true },
                SecurityID: { querystring: true }
            }
        });

        endpoint({
            id: props.ID,
            SecurityID: Config.get('SecurityID')
        })
            .then(response => response.json())
            .then(output => {
                setLoading({ complete: true, failed: !!output.error, submitting: false });
            })
            .catch(() => {
                setLoading({ complete: true, failed: true, submitting: false });
            });
    }

    if (loading.complete || loading.submitting) {
        return null;
    }

    const created = moment(props.Created);
    const lastAccessed = moment(props.LastAccessed);
    const lastAccessedElapsed = moment.utc(props.LastAccessed).fromNow();
    const currentStr = i18n._t('SessionManager.CURRENT', 'Current');
    const lastActiveStr = i18n._t('SessionManager.LAST_ACTIVE', 'last active');
    const logOutStr = i18n._t('SessionManager.LOG_OUT', 'Log Out');

    const activityTooltip = i18n.inject(
        i18n._t('Admin.ACTIVITY_TOOLTIP_TEXT', 'Signed in {signedIn}, Last active {lastActive}'),
        {
            signedIn: created.format('L LT'),
            lastActive: lastAccessed.format('L LT')
        }
    );

    return (
        <div className="login-session">
            <div>{props.UserAgent}</div>
            {props.IsCurrent &&
                <strong className={'text-success'} data-toggle="tooltip" data-placement="top" title={activityTooltip}>
                    {currentStr}
                </strong>
            }
            <div className="text-muted">
                {props.IPAddress}
                {!props.IsCurrent &&
                    <span data-toggle="tooltip" data-placement="top" title={activityTooltip}>
                        , {lastActiveStr} {lastAccessedElapsed}
                    </span>
                }
            </div>
            {!props.IsCurrent && <a
                role={'button'}
                tabIndex={'0'}
                className={'login-session__logout'}
                onClick={logOut}
            >{logOutStr}</a>}
        </div>
    );
}

LoginSession.propTypes = {
    ID: PropTypes.number.isRequired,
    IPAddress: PropTypes.string.isRequired,
    IsCurrent: PropTypes.bool,
    UserAgent: PropTypes.string,
    Created: PropTypes.string.isRequired,
    LastAccessed: PropTypes.string.isRequired,
    LogOutEndpoint: PropTypes.string.isRequired,
};

export default LoginSession;
