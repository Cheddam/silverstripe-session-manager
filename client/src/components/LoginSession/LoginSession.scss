.login-session__logout:not([href]) {
    cursor: pointer;
    color: $link-color;

    &:hover, &:active {
        text-decoration: underline;
    }
}
