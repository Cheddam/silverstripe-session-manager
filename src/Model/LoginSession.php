<?php

namespace Kinglozzer\SessionManager\Model;

use SilverStripe\Control\HTTPRequest;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\FieldType\DBDatetime;
use SilverStripe\Security\Member;
use SilverStripe\Security\RememberLoginHash;
use UAParser\Parser;

class LoginSession extends DataObject
{
    private static $db = [
        'LastAccessed' => 'DBDatetime',
        'IPAddress' => 'Varchar(45)',
        'UserAgent' => 'Text',
        'Persistent' => 'Boolean'
    ];

    private static $has_one = [
        'Member' => Member::class
    ];

    private static $belongs_to = [
        'LoginHash' => RememberLoginHash::class
    ];

    private static $indexes = [
        'LastAccessed' => true
    ];

    private static $table_name = 'LoginSession';

    private static $default_sort = 'LastAccessed DESC';

    private static $summary_fields = [
        'IPAddress' => 'IP Address',
        'LastAccessed' => 'Last Accessed',
        'Created' => 'Signed In',
        'FriendlyUserAgent' => 'User Agent'
    ];

    /**
     * The length of time a session can be inactive for before it is discarded and the
     * user is logged out
     *
     * @config
     * @var int
     */
    private static $default_session_lifetime = 3600;

    /**
     * @param Member $member
     * @param HTTPRequest $request
     * @return static|null
     */
    public static function find(Member $member, HTTPRequest $request)
    {
        $session = static::get()->filter([
            'IPAddress' => $request->getIP(),
            'UserAgent' => $request->getHeader('User-Agent'),
            'MemberID' => $member->ID,
            'Persistent' => true
        ])->first();

        return $session;
    }

    /**
     * @param Member $member
     * @param boolean $persistent
     * @param HTTPRequest $request
     * @return static
     */
    public static function generate(Member $member, $persistent = false, HTTPRequest $request)
    {
        $session = static::create()->update([
            'LastAccessed' => DBDatetime::now()->Rfc2822(),
            'IPAddress' => $request->getIP(),
            'UserAgent' => $request->getHeader('User-Agent'),
            'MemberID' => $member->ID,
            'Persistent' => $persistent
        ]);
        $session->write();

        return $session;
    }

    /**
     * @return string
     */
    public function getFriendlyUserAgent()
    {
        $parser = Parser::create();
        $result = $parser->parse($this->UserAgent);

        return sprintf('%s on %s', $result->ua->family, $result->os->toString());
    }
}
