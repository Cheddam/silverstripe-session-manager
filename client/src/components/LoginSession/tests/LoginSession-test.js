/* global jest, jasmine, describe, it, expect, beforeEach, Event */

import React from 'react';
import LoginSession from '../LoginSession';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import backend from 'lib/Backend';
import MockDate from 'mockdate';

jest.mock('@silverstripe/reactstrap-confirm', () => jest.fn().mockImplementation(
    () => Promise.resolve(true)
));

Enzyme.configure({ adapter: new Adapter() });

describe('LoginSession', () => {
    let props = null;

    beforeEach(() => {
        async function endpointFetcher() {
            return {
                error: false,
                success: true
            };
        }
        backend.createEndpointFetcher = jest.fn().mockImplementation(() => endpointFetcher);

        MockDate.set('2021-03-12 03:47:22');

        // Set window config
        window.ss.config = {
            SecurityID: 1234567890
        };

        props = {
            ID: 1,
            IPAddress: '127.0.0.1',
            UserAgent: 'Chrome on Mac OS X 10.15.7',
            Created: '2021-01-20 00:33:41',
            LastAccessed: '2021-03-11 03:47:22',
            LogOutEndpoint: 'admin/loginsession/remove',
        };
    });

    describe('render()', () => {
        it('should match the snapshot', () => {
            const wrapper = shallow(<LoginSession {...props} />);
            expect(wrapper.html()).toMatchSnapshot();
        });

        it('should log sessions out correctly', done => {
            const wrapper = shallow(<LoginSession {...props} />);
            wrapper.find('.login-session__logout').simulate('click');

            setTimeout(() => {
                expect(wrapper.html()).toMatchSnapshot();
                done();
            });
        });
    });
});
