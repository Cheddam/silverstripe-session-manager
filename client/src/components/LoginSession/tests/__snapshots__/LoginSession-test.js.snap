// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSession render() should log sessions out correctly 1`] = `"<div class=\\"login-session hidden\\"><p>Chrome on Mac OS X 10.15.7</p><p class=\\"text-muted\\">127.0.0.1<span data-toggle=\\"tooltip\\" data-placement=\\"top\\" title=\\"Signed in 01/20/2021 12:33 AM, Last active 03/11/2021 3:47 AM\\">, last active 11 hours ago...</span></p><a role=\\"button\\" tabindex=\\"0\\" class=\\"login-session__logout\\">Logging out...</a></div>"`;

exports[`LoginSession render() should match the snapshot 1`] = `"<div class=\\"login-session \\"><p>Chrome on Mac OS X 10.15.7</p><p class=\\"text-muted\\">127.0.0.1<span data-toggle=\\"tooltip\\" data-placement=\\"top\\" title=\\"Signed in 01/20/2021 12:33 AM, Last active 03/11/2021 3:47 AM\\">, last active 11 hours ago...</span></p><a role=\\"button\\" tabindex=\\"0\\" class=\\"login-session__logout\\">Log out</a></div>"`;
